### PR TITLE
Handle and log the UnicodeDecodeError

### PR DESCRIPTION
**Pull request by d-terletskyi**
Fixes ZEN-32208.

To handle the issue with UnicodeDecodeError the decoding procedure was
modified by adding additional checks. The procedure tries to decode
values using zCollectorDecoding if it is defined, otherwise, it tries to
apply the utf-8 and then latin-1encodings simultaneously logging
information about this. If decoding of the value in strict mode causes
errors, then it will be done in ignore mode, therefore non-Unicode
symbols will be ignored.

### CODE REVIEW

File: Products/DataCollector/ApplyDataMap/datamaputils.py [hunk 1]
```diff
@@ -158,26 +158,63 @@ def _decode_value(value, obj):
...
+
+            log.warn("The string %s was decoded to %s ignoring symbols "
+                     "which is impossible to decode into '%s' ",
+                     value, codec, codec)
+
+    encoder = sys.getdefaultencoding()
+    if not encoder:
+        encoder = 'utf-8'
+    value = value.encode(encoder)
+
```
Comment (jpeacock-zenoss) [inline Products/DataCollector/ApplyDataMap/datamaputils.py:212]: Now that I've had some time to think about this issue and review your changes in response to our original conversation, I think a different strategy will have some benefits.

After relearning that the `decode` method has an `errors` argument, I think the strategy of trying to guess the encoding is more work than necessary to accomplish our goal.  The goal of the `_decode_value` function is to normalize strings to the system's default encoding.  For troubleshooting purposes, a side goal would be to log when specifying `zCollectorDecoding` would be helpful for decoding the value.  Another thing to note is that we've configured `sys.getdefaultencoding` to return "utf-8" as the default codec (have a look at /opt/zenoss/lib/python2.7/sitecustomize.py in an image).

So, maybe an implementation like the following will suffice:
```
    try:
        value = value.decode(codec)
    except UnicodeDecodeError as ex:
        value = value.decode(codec, errors="ignore")
        log.warn(
            "Unable to decode string using codec '%s'.  "
            "Using a modified string without errors.  "
            "Please set zCollectorDecoding to the correct codec. "
            "object=%r modified-value='%s' error=%s",
            codec, obj, value, ex,
        )
            
    return value.encode(sys.getdefaultencoding())
```

**Review by jpeacock-zenoss — CHANGES_REQUESTED** [review]

File: Products/DataCollector/ApplyDataMap/datamaputils.py [hunk 2]
```diff
@@ -158,26 +158,63 @@ def _decode_value(value, obj):
...
         codec = sys.getdefaultencoding()
 
-    value = value.decode(codec)
-    value = value.encode(sys.getdefaultencoding())
+    if not codec:
+        log.warn(
+            "zCollectorDecoding is not defined! It can cause problems "
+            "with the modeling and even can crash it, therefore specify "
+            "the correct data encoding as a value of "
+            "zCollectorDecoding.")
```
[the remark this answers is not in the document]
Comment (d-terletskyi) [reply Products/DataCollector/ApplyDataMap/datamaputils.py:180]: Understood

File: Products/DataCollector/ApplyDataMap/datamaputils.py [hunk 3]
```diff
@@ -158,26 +158,63 @@ def _decode_value(value, obj):
...
+
+        obj.zCollectorDecoding = 'utf-8'
+        codec = obj.zCollectorDecoding
```
[the remark this answers is not in the document]
Comment (d-terletskyi) [reply Products/DataCollector/ApplyDataMap/datamaputils.py:183]: Understood.

File: Products/DataCollector/ApplyDataMap/datamaputils.py [hunk 4]
```diff
@@ -158,26 +158,63 @@ def _decode_value(value, obj):
...
+
+    try:
+        value = value.decode(codec)
+    except UnicodeDecodeError as ex1:
+        log.warn("The string %s cannot be correctly decoded using '%s' and "
+                 "causes error: %s! Please verify and specify the correct "
+                 "data encoding as a value of zCollectorDecoding.", value,
+                 codec, ex1)
```
[the remark this answers is not in the document]
Comment (d-terletskyi) [reply Products/DataCollector/ApplyDataMap/datamaputils.py:191]: Do you mean we need to write into the log only one final message after the decoding/encoding resolution?

[the remark this answers is not in the document]
Comment (jpeacock-zenoss) [reply Products/DataCollector/ApplyDataMap/datamaputils.py:191]: Yes.  Like in the example code I wrote below.

**Comment by d-terletskyi** [comment]
The fix was improved according to all remarks and suggestions noted above.

**Review by jpeacock-zenoss — CHANGES_REQUESTED** [review]

**Comment by d-terletskyi** [comment]
The text of the warning was slightly modified, according to the remark noted above.

**Review by jpeacock-zenoss — APPROVED** [review]

**Pull request merged**